### PR TITLE
Renamed DictionaryLiteral to KeyValuePairs

### PR DESCRIPTION
**Pull request by masters3d**
Implementation for https://github.com/apple/swift-evolution/pull/850

"There is no strong motivation to deprecate. The type does not produce
active harm. Instead, it adds measurable (if small) utility and will be
part of the ABI. A sensible renaming mitigates the most problematic
issue with the type."

### CODE REVIEW

**Comment by masters3d** [comment]
@swift-ci Please smoke test

**Comment by nh7a** [comment]
The proposal is suggesting `KeyValueList`, not `KeyValueArray`, though.

File: include/swift/AST/Expr.h
Comment (benrimmington) [inline include/swift/AST/Expr.h]: I think this refers to `DictionaryExpr` (which is a subclass of `CollectionExpr`).

**Comment by benrimmington** [comment]
`DictionaryLiteral` is used in:

* [benchmark/single-source/RomanNumbers.swift](https://github.com/apple/swift/blob/master/benchmark/single-source/RomanNumbers.swift)
* [benchmark/single-source/StringMatch.swift](https://github.com/apple/swift/blob/master/benchmark/single-source/StringMatch.swift)
* [stdlib/private/StdlibUnittest/TypeIndexed.swift](https://github.com/apple/swift/blob/master/stdlib/private/StdlibUnittest/TypeIndexed.swift)
* [stdlib/public/core/CompilerProtocols.swift](https://github.com/apple/swift/blob/b6e637e03d012121c59c02aab7915f563401814e/stdlib/public/core/CompilerProtocols.swift#L665-L668)
* [test/stdlib/DictionaryLiteral.swift](https://github.com/apple/swift/blob/master/test/stdlib/DictionaryLiteral.swift)

File: include/swift/AST/Expr.h
Comment (benrimmington) [reply include/swift/AST/Expr.h]: The tuples are probably those given to any [`init(dictionaryLiteral:)`](https://github.com/apple/swift/blob/b6e637e03d012121c59c02aab7915f563401814e/stdlib/public/core/CompilerProtocols.swift#L699-L706) initializer. I don't think `AST/Expr.h` would need to have any knowledge of the `KeyValueArray` type.

Comment (masters3d) [reply include/swift/AST/Expr.h]: This is part of the confusion and reason behind the rename. I believe while `AST/Expr.h` has not much to do with concrete types, the comment seems to be referring to the concrete type. Since  this comment is not user facing I am not too worried about it.  You are welcome to change it back. I added you as a collaborator to my branch. Are you able to start the CI? Thanks!

Comment (benrimmington) [reply include/swift/AST/Expr.h]: I haven't got commit access, so I'm not able to trigger CI testing.

But the [swift-package-manager](https://github.com/apple/swift-package-manager/search?q=DictionaryLiteral&type=Code) didn't compile last time, did it?

You'll either need to:

- use [cross repository testing](https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#cross-repository-testing) with another PR, or
- wait until the compiler supports typealias extensions.

I also wondered if a [generic typealias](https://github.com/apple/swift-evolution/blob/master/proposals/0048-generic-typealias.md) might help, but it didn't seem to (with a recent Swift 4.2 snapshot).

**Comment by masters3d** [comment]
@swift-ci Please test source compatibility

**Comment by aciidb0mb3r** [comment]
@swift-ci Please test source compatibility

**Comment by aciidb0mb3r** [comment]
@swift-ci please test source compatibility

**Comment by masters3d** [comment]
none of the failures I saw are related https://ci.swift.org/job/swift-PR-source-compat-suite/1219/artifact/swift-source-compat-suite/

Aside from the Package Manager, most people do not seem to know or extend the type so I think they will not be any source compatibility. 
https://github.com/search?p=2&q="extension+DictionaryLiteral+where"&type=Code&utf8=✓

**Comment by masters3d** [comment]
Updated. Not sure what to do about [stdlib-stable.json](https://github.com/apple/swift/blob/master/test/api-digester/stdlib-stable.json

**Comment by lattner** [comment]
Great works @masters3d.  The patch LGTM, but it would be good to get a standard library reviewer like @airspeedswift or @dabrahams to take a look.

**Review by airspeedswift — APPROVED** [review]
LGTM with the exception of the question of the deprecation.

**Comment by airspeedswift** [comment]
This all looks good to me. However, I am not sure about the deprecation in 4.2. While we could do it in 4.2, I think that's a little aggressive.

My preference would be to add the new name as a typealias in 4.2 now, and to deprecate the old name in 5.0. That way, people will be able to update their code bases in advance, and once 5.0 comes around, we can deprecate the old name in a way that means users will be able to compile without warnings with both the 4.2 and 5.0 compiler.

**Comment by lattner** [comment]
I don't think this type is widely used (does it appear in the source compat suite at all?) so I don't think that warning in 4.2 will be too aggressive.  That said, your approach is also totally fine with me.

File: CHANGELOG.md
Comment (benrimmington) [inline CHANGELOG.md]: You'll need to add a reference to the end of this file. The missing references are:

```
[SE-0213]: <https://github.com/apple/swift-evolution/blob/master/proposals/0213-literal-init-via-coercion.md>
[SE-0214]: <https://github.com/apple/swift-evolution/blob/master/proposals/0214-DictionaryLiteral.md>
[SE-0215]: <https://github.com/apple/swift-evolution/blob/master/proposals/0215-conform-never-to-hashable-and-equatable.md>
[SE-0216]: <https://github.com/apple/swift-evolution/blob/master/proposals/0216-dynamic-callable.md>
```

**Comment by benrimmington** [comment]
There's a documentation comment to update in:

* [stdlib/public/core/CompilerProtocols.swift](https://github.com/apple/swift/blob/50fb3b8496f9501c95326a1f9e33638e56797fb0/stdlib/public/core/CompilerProtocols.swift#L665-L668)

The `DictionaryLiteral` typealias will be used in:

* [benchmark/single-source/RomanNumbers.swift](https://github.com/apple/swift/blob/50fb3b8496f9501c95326a1f9e33638e56797fb0/benchmark/single-source/RomanNumbers.swift#L28)
* [stdlib/private/StdlibUnittest/TypeIndexed.swift](https://github.com/apple/swift/blob/50fb3b8496f9501c95326a1f9e33638e56797fb0/stdlib/private/StdlibUnittest/TypeIndexed.swift#L101)

**Comment by masters3d** [comment]
Updated.

File: CHANGELOG.md
Comment (masters3d) [reply CHANGELOG.md]: The others are not mention on the document yet so I only added [SE-0213]

**Comment by masters3d** [comment]
Can this be merged?

**Comment by airspeedswift** [comment]
@swift-ci please test and merge

**Comment by airspeedswift** [comment]
Hi @masters3d can you rebase this PR and we can try again? Thanks!

**Comment by masters3d** [comment]
done.

**Comment by airspeedswift** [comment]
@swift-ci please test

File: CHANGELOG.md
Comment (benrimmington) [reply CHANGELOG.md]: This changelog entry will probably need to be moved to the Swift 5.0 section.

The other typealias from #17711 could be documented here (Swift 4.2) instead.

File: stdlib/public/core/Mirror.swift
Comment (benrimmington) [inline stdlib/public/core/Mirror.swift]: Similar to #17711, should this be a generic typealias?

```diff
 @available(swift, deprecated: 5.0, renamed: "KeyValuePairs")
-public typealias DictionaryLiteral = KeyValuePairs
+public typealias DictionaryLiteral<Key, Value> = KeyValuePairs<Key, Value>
```

Comment (masters3d) [reply stdlib/public/core/Mirror.swift]: I don't have a preference either way. @benrimmington you have write access to the branch. Please feel free to update it.

Comment (benrimmington) [reply stdlib/public/core/Mirror.swift]: @masters3d I've rebased and tested the changes locally, but I can't push to your branch. Please try adding me as a collaborator again.

**Comment by masters3d** [comment]
It looks like you need to accept the invitation to collaborate.
https://github.com/masters3d/swift/invitations

File: stdlib/public/core/Mirror.swift
Comment (benrimmington) [reply stdlib/public/core/Mirror.swift]: @masters3d I'm probably doing something wrong, but authentication fails on the command line (`git push masters3d +renamed-to-KeyValueArray`).

I also tried pushing from Xcode, but it fails because your branch is 1105 commits behind `apple:master`. If you can rebase your branch from upstream, I'll try again.

Comment (masters3d) [reply stdlib/public/core/Mirror.swift]: https://help.github.com/articles/pushing-to-a-remote/image

You should be able to push, try pushing a branch with a new name (that should work)

Comment (benrimmington) [reply stdlib/public/core/Mirror.swift]: @masters3d If I create a new branch in your repository, it won't update this pull request. I've pushed to [`benrimmington:renamed-to-KeyValueArray`](https://github.com/benrimmington/swift/tree/renamed-to-KeyValueArray) for now, but it will be simpler to ignore that and make your own changes instead.

**Comment by masters3d** [comment]
Thanks @benrimmington I reset my branch to yours.

**Comment by benrimmington** [comment]
@airspeedswift Please could you trigger swift-ci testing again.

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by benrimmington** [comment]
@airspeedswift The smoke test has finished. Can this be merged? There have been a few extra commits, since you last approved this.

**Comment by benrimmington** [comment]
@masters3d Please reset your branch to [`benrimmington:renamed-to-KeyValueArray`](https://github.com/benrimmington/swift/tree/renamed-to-KeyValueArray), to resolve another conflict.

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please clean test

**Comment by benrimmington** [comment]
The failure is due to a missing type change in [test/api-digester/source-stability.swift.expected](https://github.com/apple/swift/blob/master/test/api-digester/source-stability.swift.expected)

```diff
 Var Dictionary.values has declared type change from LazyMapCollection<[Key : Value], Value> to Dictionary<Key, Value>.Values
+Constructor Mirror.init(_:children:displayStyle:ancestorRepresentation:) has parameter 1 type change from DictionaryLiteral<String, Any> to KeyValuePairs<String, Any>
 Constructor String.init(_:) has return type change from String? to String
```

**Comment by airspeedswift** [comment]
@akyrtzi should that api difference be firing? the one is a typealias of the other.

**Comment by airspeedswift** [comment]
I mean, it'd be an ABI break once we stabilize the ABI, but it shouldn't be a source compatibility break.

**Comment by nkcsgexi** [comment]
Can we update the expected change list for now to unblock the merging of this patch?

**Comment by airspeedswift** [comment]
sure, just want to check if it's a bug that it flagged.

**Comment by nkcsgexi** [comment]
yeah, i think the api checker currently isn't smart enough to tell a type alias change is benign for source compatibility. This won't be an issue when we check abi stability because abi checker compares module dumps with all types canonicalized.

**Comment by airspeedswift** [comment]
Hi @masters3d can you add that entry to the expected results to fix the test for now? Thanks!

**Comment by benrimmington** [comment]
@masters3d I've updated [`benrimmington:renamed-to-KeyValueArray`](https://github.com/benrimmington/swift/tree/renamed-to-KeyValueArray) again, if you'd prefer to reset your branch.

**Comment by masters3d** [comment]
@airspeedswift @benrimmington updated

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by benrimmington** [comment]
@airspeedswift You'll probably need a "clean test", because my branch had been rebased.

@masters3d I've rebased again for a merge conflict, but it looks like you've already fixed it?

**Comment by airspeedswift** [comment]
Those failures are spurious CI artifacts, not actual failures.

**Comment by airspeedswift** [comment]
@swift-ci please test macOS platform

**Comment by airspeedswift** [comment]
@shahmishal any idea why macOS tests aren't triggering on this PR?

**Comment by shahmishal** [comment]
It's running - https://ci.swift.org/job/swift-PR-osx/7078/

**Comment by benrimmington** [comment]
Lines 478...480 of Mirror.swift could be re-indented (in a follow-up PR).

benrimmington/swift@8aa413ccbceef56faeaf3d60ef5a3132af1e2581

**Comment by masters3d** [comment]
Thank you swift team! Thanks @benrimmington !

**Pull request merged**
